src/sdam/state/mod.rs

```diff
@@ -514,8 +583,46 @@ impl TopologyState {
         topology: WeakTopology,
     ) -> std::result::Result<Option<TopologyDescriptionDiff>, String> {
         let old_description = self.description.clone();
+
+        if let Some(ref handler) = options.sdam_event_handler {
```

Comment:
The server changed events need to be moved into the diff block as well I think, since other servers in the topology may get updated as part of the new description. An example is if we have a `ReplicaSetWithPrimary` topology and we get a server description from a server that says it's the primary, we need to mark the server we think is a primary as unknown. To implement this I think we'll need to update `TopologyDescription::diff` to check each server for equality and report the changed ones in another field similar to the other two.